File: 200.number-of-islands.cpp
/*
 * @lc app=leetcode id=200 lang=cpp
 *
 * [200] Number of Islands
 */

#include <queue>
#include <vector>
#include <utility>

using std::queue;
using std::pair;
using std::vector;

// @lc code=start
class Solution {
    static constexpr int DIRN[] {-1, 0, 1, 0, -1};

public:
    // BFS solution
    // T: O(m * n), m := the number of rows, n := the number of columns
    // S: O(m * n)
    int numIslands(vector<vector<char>>& grid) {
        int ans = 0;
        queue<pair<int, int>> q;
        for (int i = 0; i < grid.size(); ++i) {
            for (int j = 0; j < grid[i].size(); ++j) {
                if (grid[i][j] == '1') {
                    // q should be empty when a new island is just found
                    ++ans;
                    grid[i][j] = '0';   // mark visited

                    // For each island, do BFS and mark all cells of that current island visited
                    q.emplace(i, j);
                    while (!q.empty()) {
                        int qs = q.size();
                        while (qs-- > 0) {
                            const auto p = q.front(); q.pop();
                            for (int d = 0; d < 4; ++d) {
                                int ni = p.first + DIRN[d],
                                    nj = p.second + DIRN[d + 1];
                                if (ni >= 0 && ni < grid.size() && nj >= 0 && nj < grid[ni].size() && grid[ni][nj] == '1') {
                                    grid[ni][nj] = '0';
                                    q.emplace(ni, nj);
                                }
                            }
                        }
                    }
                }
            }
        }
        return ans;
    }

    /*
    // DFS solution
    // T: O(m * n), m := the number of rows, n := the number of columns
    // S: O(m * n)
    int numIslands(vector<vector<char>>& grid) {
        int ans = 0;
        for (int i = 0; i < grid.size(); ++i) {
            for (int j = 0; j < grid[i].size(); ++j) {
                if (grid[i][j] == '1') {
                    ++ans;
                    dfs(grid, i, j);
                }
            }
        }
        return ans;
    }

private:
    static void dfs(vector<vector<char>>& grid, int i, int j) {
        grid[i][j] = '0';   // marked visited
        for (int d = 0; d < 4; ++d) {
            int ni = i + DIRN[d], nj = j + DIRN[d + 1];
            if (ni >= 0 && ni < grid.size() && nj >= 0 && nj < grid[ni].size() && grid[ni][nj] == '1') {
                dfs(grid, ni, nj);
            }
        }
    }
    */
};
// @lc code=end

File: 200.number-of-islands.go
/*
 * @lc app=leetcode id=200 lang=golang
 *
 * [200] Number of Islands
 */
package lc

// @lc code=start
var dirn = [5]int{1, 0, -1, 0, 1}

/*
// Perform BFS for each unvisited island cell
type Pos struct {
	r, c int
}

// T: O(m * n)
// S: O(m * n)
func numIslands(grid [][]byte) int {
	m := len(grid)
	n := len(grid[0])

	q := make(chan *Pos, m*n)
	ans := 0

	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if grid[i][j] == '1' {
				ans++

				q <- &Pos{i, j}
				grid[i][j] = '0'

				for len(q) != 0 {
					pos := <-q

					for d := 0; d < 4; d++ {
						nr := pos.r + dirn[d]
						nc := pos.c + dirn[d+1]
						if 0 <= nr && nr < m && 0 <= nc && nc < n && grid[nr][nc] == '1' {
							q <- &Pos{nr, nc}
							grid[nr][nc] = '0'
						}
					}
				}
			}
		}
	}
	return ans
}
*/

// DFS
func dfs(grid *[][]byte, r int, c int) {
	if r < 0 || r >= len(*grid) || c < 0 || c >= len((*grid)[r]) || (*grid)[r][c] == '0' {
		return
	}
	(*grid)[r][c] = '0'
	for d := 0; d < 4; d++ {
		dfs(grid, r+dirn[d], c+dirn[d+1])
	}
}

// T: O(m * n)
// S: O(m * n) recursion stack
func numIslands(grid [][]byte) int {
	ans := 0
	for i := range grid {
		for j := range grid[i] {
			if grid[i][j] == '1' {
				ans++
				dfs(&grid, i, j)
			}
		}
	}
	return ans
}

// @lc code=end
